src/internal/connector/discovery/api/users.go

```diff
@@ -165,6 +166,78 @@ func (c Users) GetByID(ctx context.Context, identifier string) (models.Userable,
 	return resp, err
 }
 
+// GetUserInfo provide information about a user like- userPurpose/type,
+// if mailbox and onedrive is present for user.
+func (c Users) GetUserInfo(ctx context.Context, userID string) (string, bool, bool, string, error) {
+	var (
+		rawURL                         = "https://graph.microsoft.com/v1.0/users/" + userID + "/mailboxSettings"
+		apadtor                        = c.stable.Adapter()
+		builder                        = users.NewUserItemRequestBuilder(rawURL, apadtor)
+		hasMailBox, hasOneDrive, ok    bool
+		userPurpose, errGetUserPurpose string
+		err                            error
+	)
+
+	// verify mailbox enabled for user
+	_, err = c.stable.Client().UsersById(userID).MailFolders().Get(ctx, nil)
+	hasMailBox = true
+
+	if err != nil {
+		if !graph.IsErrExchangeMailFolderNotFound(err) {
+			return "", false, false, "", graph.Wrap(ctx, err, "error getting mail folder")
+		}
+
+		logger.Ctx(ctx).Infof("resource owner does not have a mailbox enabled")
+
+		hasMailBox = false
+	}
+
+	// verify onedrive enabled for user
+	_, err = c.stable.Client().UsersById(userID).Drives().Get(ctx, nil)
+	hasOneDrive = true
+
+	if err != nil {
+		err = graph.Stack(ctx, err)
+		if !clues.HasLabel(err, graph.LabelsMysiteNotFound) {
+			return "", false, false, "", graph.Wrap(ctx, err, "getting onedrive")
+		}
+
+		logger.Ctx(ctx).Infof("resource owner does not have a drive")
+
+		hasOneDrive = false
+	}
+
+	// get userPurpose for user
+	newItem, err := builder.Get(ctx, nil)
```

Comment:
`builder` isn't getting re-used.  Rather than separating the builder creation from its usage, both should occur here in the same line.

Also, stepping back from the granular implementation a bit.  This function is a bit overloaded, and is going to be more difficult than necessary to write tests for.  Consider a layout like the following as an alternative, which breaks each call out to its own func and only uses this func to control aggregation of the results:

```go
func (c Users) GetInfo(ctx) (UserInfo, error) {
  userInfo := newUserInfo()
  
  userInfo.Exchange, err = c.allowsExchange(ctx)
  if err { return }
  
  userInfo.OneDrive, err = c.allowsOneDrive(ctx)
  if err { return }
  
  userInfo.purpose, err = c.getPurpose(ctx)
  if err { return }
  
  return userInfo, nil
}

func (c Users) allowsExchange(ctx) (bool, err) {...}
func (c Users) allowsOneDrive(ctx) (bool, err) {...}
func (c Users) getPurpose(ctx) (string, err) {...}
```

This does the following:
1. makes the GetInfo func easy to read and grok.
2. places all the granular handling into self-contained functions.  funcs like that are easy to mock and test and make clear assertions about without cross contamination from other calls.
3. provides a set of re-usable funcs in case we want to make the same individual calls elsewhere.